File: src/tools/registry.ts
// 工具注册表：每个工具 = 名称 + 描述 + 参数 JSON Schema + handler。
// LLM 只看得到前三样（拼进 system prompt），runtime 负责校验参数再调 handler。

export interface JsonSchema {
  type: "object";
  properties: Record<string, { type: string; description?: string; enum?: string[] }>;
  required?: string[];
}

export interface ToolContext {
  /** 当前 session 的可变状态袋，有状态工具（todo）把数据放这里，天然按 session 隔离 */
  sessionState: Record<string, unknown>;
  userId: string;
  sessionId: string;
  /** 当前轮次（runtime 传入；remember 用它记 source，#19 Q3）。直接 invoke 时可不给 */
  turn?: number;
}

export interface ToolDefinition {
  name: string;
  description: string;
  parameters: JsonSchema;
  handler: (args: Record<string, any>, ctx: ToolContext) => Promise<string>;
  /**
   * 结果精简：工具原始输出可能很长，塞回 context 前先过这一道。
   * 不提供则按默认截断。
   */
  compact?: (raw: string) => string;
  /**
   * 进 trace 前对 args 脱敏（白名单落盘）：用户内容类参数只留长度 / 摘要。
   * 不提供则 args 原样进 trace。只影响 trace，不影响 handler 收到的参数。
   */
  redact?: (args: Record<string, unknown>) => Record<string, unknown>;
}

export interface ToolSpec {
  name: string;
  description: string;
  parameters: JsonSchema;
}

export interface ToolResult {
  ok: boolean;
  content: string;
  durationMs: number;
}

const DEFAULT_MAX_CHARS = 1500;

export class ToolRegistry {
  private tools = new Map<string, ToolDefinition>();

  register(def: ToolDefinition): this {
    if (this.tools.has(def.name)) throw new Error(`工具重名：${def.name}`);
    this.tools.set(def.name, def);
    return this;
  }

  has(name: string): boolean {
    return this.tools.has(name);
  }

  /** trace 用：按工具声明脱敏后的 args；未注册或未声明 redact 则原样 */
  redact(name: string, args: Record<string, unknown>): Record<string, unknown> {
    const def = this.tools.get(name);
    return def?.redact ? def.redact(args) : args;
  }

  specs(): ToolSpec[] {
    return [...this.tools.values()].map(({ name, description, parameters }) => ({ name, description, parameters }));
  }

  /** 校验参数 → 执行 → 精简。任何失败都变成 ok:false 的结果，不向上抛。 */
  async invoke(name: string, args: Record<string, unknown>, ctx: Partial<ToolContext>): Promise<ToolResult> {
    const started = Date.now();
    const done = (ok: boolean, content: string): ToolResult => ({ ok, content, durationMs: Date.now() - started });

    const def = this.tools.get(name);
    if (!def) return done(false, `工具 "${name}" 未注册。可用工具：${[...this.tools.keys()].join(", ")}`);

    const problem = validateArgs(def.parameters, args);
    if (problem) return done(false, `参数错误：${problem}`);

    const fullCtx: ToolContext = {
      sessionState: ctx.sessionState ?? {},
      userId: ctx.userId ?? "anon",
      sessionId: ctx.sessionId ?? "default",
      turn: ctx.turn,
    };
    try {
      const raw = await def.handler(args, fullCtx);
      const compacted = def.compact ? def.compact(raw) : truncate(raw, DEFAULT_MAX_CHARS);
      return done(true, compacted);
    } catch (e) {
      return done(false, `工具执行失败：${(e as Error).message}`);
    }
  }
}

/** 极简 JSON Schema 校验：只管 required 与顶层类型，够这题用；返回第一条问题描述。 */
export function validateArgs(schema: JsonSchema, args: Record<string, unknown>): string | null {
  for (const key of schema.required ?? []) {
    if (args[key] === undefined || args[key] === null) return `缺少必填参数 "${key}"`;
  }
  for (const [key, val] of Object.entries(args)) {
    const prop = schema.properties[key];
    if (!prop) return `未知参数 "${key}"，允许的参数：${Object.keys(schema.properties).join(", ")}`;
    const actual = Array.isArray(val) ? "array" : typeof val;
    const expected = prop.type === "integer" ? "number" : prop.type;
    if (actual !== expected) return `参数 "${key}" 应为 ${prop.type}，实际是 ${actual}`;
    if (prop.enum && !prop.enum.includes(String(val))) return `参数 "${key}" 只能是 ${prop.enum.join(" / ")}`;
  }
  return null;
}

export function truncate(s: string, max: number): string {
  return s.length <= max ? s : s.slice(0, max) + `\n…（已截断，原文 ${s.length} 字符）`;
}

File: mini_agent/tools/registry.py
"""工具注册表：每个工具 = 名称 + 描述 + 参数 JSON Schema + handler。
LLM 只看得到前三样（拼进 system prompt），runtime 负责校验参数再调 handler。

ToolContext（dict）：{sessionState, userId, sessionId, turn?}
    sessionState  当前 session 的可变状态袋，有状态工具（todo）把数据放这里，天然按 session 隔离
    turn          当前轮次（runtime 传入；remember 用它记 source，#19 Q3）。直接 invoke 时可不给
ToolResult（dict）：{ok, content, durationMs}
"""
import time
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional

JsonSchema = Dict[str, Any]
Handler = Callable[[Dict[str, Any], Dict[str, Any]], str]


@dataclass
class ToolDefinition:
    name: str
    description: str
    parameters: JsonSchema
    handler: Handler
    # 结果精简：工具原始输出可能很长，塞回 context 前先过这一道。不提供则按默认截断。
    compact: Optional[Callable[[str], str]] = None
    # 进 trace 前对 args 脱敏（白名单落盘）：用户内容类参数只留长度 / 摘要。不提供则 args 原样进 trace。只影响 trace，不影响 handler 收到的参数。
    redact: Optional[Callable[[Dict[str, Any]], Dict[str, Any]]] = None

    def spec(self) -> dict:
        return {"name": self.name, "description": self.description, "parameters": self.parameters}


DEFAULT_MAX_CHARS = 1500


class ToolRegistry:
    def __init__(self) -> None:
        self._tools: Dict[str, ToolDefinition] = {}

    def register(self, d: ToolDefinition) -> "ToolRegistry":
        if d.name in self._tools:
            raise ValueError(f"工具重名：{d.name}")
        self._tools[d.name] = d
        return self

    def has(self, name: str) -> bool:
        return name in self._tools

    def redact(self, name: str, args: Dict[str, Any]) -> Dict[str, Any]:
        """trace 用：按工具声明脱敏后的 args；未注册或未声明 redact 则原样"""
        d = self._tools.get(name)
        return d.redact(args) if d and d.redact else args

    def specs(self) -> List[dict]:
        return [d.spec() for d in self._tools.values()]

    def invoke(self, name: str, args: Dict[str, Any], ctx: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """校验参数 → 执行 → 精简。任何失败都变成 ok:false 的结果，不向上抛。"""
        started = time.monotonic()

        def done(ok: bool, content: str) -> Dict[str, Any]:
            return {"ok": ok, "content": content, "durationMs": int((time.monotonic() - started) * 1000)}

        d = self._tools.get(name)
        if d is None:
            return done(False, f'工具 "{name}" 未注册。可用工具：{", ".join(self._tools.keys())}')
        problem = validate_args(d.parameters, args)
        if problem:
            return done(False, f"参数错误：{problem}")
        ctx = ctx or {}
        full_ctx = {
            "sessionState": ctx.get("sessionState") if ctx.get("sessionState") is not None else {},
            "userId": ctx.get("userId") or "anon",
            "sessionId": ctx.get("sessionId") or "default",
            "turn": ctx.get("turn"),
        }
        try:
            raw = d.handler(args, full_ctx)
            return done(True, d.compact(raw) if d.compact else truncate(raw, DEFAULT_MAX_CHARS))
        except Exception as e:  # noqa: BLE001 —— 工具的任何异常都以结果形式回喂
            return done(False, f"工具执行失败：{e}")


def _js_typeof(v: Any) -> str:
    if isinstance(v, bool):
        return "boolean"
    if isinstance(v, (int, float)):
        return "number"
    if isinstance(v, str):
        return "string"
    if isinstance(v, list):
        return "array"
    return "object"


def validate_args(schema: JsonSchema, args: Dict[str, Any]) -> Optional[str]:
    """极简 JSON Schema 校验：只管 required 与顶层类型，够这题用；返回第一条问题描述。"""
    for key in schema.get("required", []):
        if args.get(key) is None:
            return f'缺少必填参数 "{key}"'
    props = schema.get("properties", {})
    for key, val in args.items():
        prop = props.get(key)
        if prop is None:
            return f'未知参数 "{key}"，允许的参数：{", ".join(props.keys())}'
        actual = _js_typeof(val)
        expected = "number" if prop["type"] == "integer" else prop["type"]
        if actual != expected:
            return f'参数 "{key}" 应为 {prop["type"]}，实际是 {actual}'
        if prop.get("enum") and str(val) not in prop["enum"]:
            return f'参数 "{key}" 只能是 {" / ".join(prop["enum"])}'
    return None


def truncate(s: str, max_chars: int) -> str:
    return s if len(s) <= max_chars else s[:max_chars] + f"\n…（已截断，原文 {len(s)} 字符）"
